include/lbann/layers/misc/dist_embedding.hpp

```diff
@@ -0,0 +1,369 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_LAYERS_MISC_DIST_EMBEDDING_HPP_INCLUDED
+#define LBANN_LAYERS_MISC_DIST_EMBEDDING_HPP_INCLUDED
+
+#include "lbann/layers/data_type_layer.hpp"
+#include "lbann/models/model.hpp"
+#include "lbann/optimizers/sgd.hpp"
+#include "lbann/utils/memory.hpp"
+
+namespace lbann {
+
+namespace dist_embedding_layer_impl {
+
+  /** Metadata for an embedding vector from a remote process. */
+  struct vector_metadata {
+    size_t source_rank{0};
+    size_t source_index{0};
+    size_t target_rank{0};
+    size_t target_index{0};
+    bool is_active{false};
+  };
```

Comment:
1. I want the struct to be available to available to CUDA kernels in an anonymous namespace. The alternative is to put the CUDA kernels in the header file (either as member functions or friends).
2. I like making zero the default value so I can reset a buffer with a memset. It's not a big deal though for the ranks and indices, since they're only read if `is_active` is `true`.

Reply:
1. Ok, then I would put them inside the struct (they're still impl details of this class), but make them public. We have a bad habit of making things globally available that, at the very least, don't need to be. It makes things look cluttered. Yes, our code is open-source, so anyone can do anything (up to whatever limits are in the license, whatever), but we can at least discourage them from doing bad things.
2. I'm not sure this is going to be faster than anything else (`operator=`); in any case, I care a lot less about the default values if the struct is tucked inside the class.